### DISCUSSION TITLE
import EDAX *.h5 file

### DISCUSSION THREAD

**Discussion by NikitaSakharov** [mtex-toolbox/mtex#1994] [Ask Anything]
Hi everyone!

Can anyone help me with importing EBSD data, which are stored in EDAX *.h5 file? All I could do is to import data from it using 
https://myhdf5.hdfgroup.org/ But imported data are in json format, so it cannot be read by MTEX.

Sincerely, Nikita.

**Comment by NikitaSakharov** [comment]
By the way. 
json file, imported from *.h5, has the following format
[[v1, v2. v3, v4, v5, v6, v7, v8, v9], c1, c2, c3, c4, c5, c6, c7, c8, c9, c10]
All the values v1-v9 are less then 1, some of them are negative. The values c3, c8 and c9 are zeroes or ones.
I think some of values among v1-v9 are Euler angles, and some are X and Y coordinates of map. How can I find out which of them are angles, which of them are coordinates?

**Comment by kilir** [comment]
Hi Nikita,
just to confirm, you tried `loadEBSD_h5` and you get an error?
Cheers,
Rüdiger

**Reply by NikitaSakharov** [reply]
Hello Rudiger!

Yes, I tried loadEBSD_h5 also and it gives me: 
Output argument "ebsd" (and maybe others) not assigned during call to "loadEBSD_h5".

**Reply by NikitaSakharov** [reply]
By the way, here is the link for file I have problem with 
https://disk.yandex.ru/d/3-192aHXMioliA

**Comment by kilir** [comment]
Hi Nikita,
I found some time to look into it.
Here's a very very rudimentary loader for the data, so that's a start.

[loadEBSD_EDAXh5.zip](https://github.com/mtex-toolbox/mtex/files/13462355/loadEBSD_EDAXh5.zip)

which you could use like any other:
```
ebsd = loadEBSD_EDAXh5('Copper test.h5','dataset',1)
```

There are probably quite many things not correct so if you have any information on the file format, it would be helpful.

Some open ends:
 - there's no info on the orientation convention, orientations come as 3x3 matrices
 - generation of x,y coordiantes could be done quite better, there are no x,y coordaintes in the file, just number of rows/columns   and stepsizes
- the unitcell is somewhat too small (rounding of steps?) and for now it's just multiplied by some number
- which convention for CS? same as .ang from edax ? 
- could it be the second map driftet? or did my x,y coordiantes mess up?

![copper](https://github.com/mtex-toolbox/mtex/assets/11697557/2ffe8eb9-38cd-4b39-a797-19b3482978d2)

Hope this is a start to improve it.

Cheers,
Rüdiger

**Reply by NikitaSakharov** [reply]
Hello Rudiger!

I've tried loadEBSD_EDAXh5 you sent, but it still does not work unfortunately :-( 
It seems to me h5 file format is a container file, which contains many many SEM data (SEM photos, spectras, EBSD data etc.). 

One of my students can take EBSD data from h5 file in ang format so now we use them to analyze EBSD data.

**Reply by NikitaSakharov** [reply]
And by the way, map really drifted in some cases. And did not in other cases. 

Frankly speaking all these EDAX soft we used to gather EBSD data was very very buggy.

Sincerely, Nikita

**Reply by kilir** [reply]
I used `loadEBSD_EDAXh5()` to successfully load the copper data you supplied without problems. So, you may have to elaborate what did not work.

Cheers,
Rüdiger

**Reply by NikitaSakharov** [reply]
Here is my code (it is the same as yours )
loadEBSD_EDAXh5('D:\[directory with h5 file]\Copper test.h5','dataset',1)

And here is what I get

ans =

    'Formula'


ans =

    'HKL Families'


ans =

    'Info'


ans =

    'Lattice Constant A'


ans =

    'Lattice Constant Alpha'


ans =

    'Lattice Constant B'


ans =

    'Lattice Constant Beta'


ans =

    'Lattice Constant C'


ans =

    'Lattice Constant Gamma'


ans =

    'Laue Group'


ans =

    'Material Name'


ans =

    'NumberFamilies'


ans =

    'Point Group'


ans =

    'Symmetry'

Index in position 2 exceeds array bounds (must not exceed 1).

Error in rotation/subsref (line 6)
    r.a = subsref(r.a,s(1));

Error in EBSD (line 114)
          pos = vector3d.byXYZ(pos(:,1),pos(:,2),0);

Error in loadEBSD_EDAXh5 (line 91)
ebsd = EBSD(rot,ebsddata.Phase,CS,ebsddata,'unitCell',uC*1.2);

**Reply by kilir** [reply]
Hi Nikita,
yes, please forget about the spamming output, comment line 51. The error stems from the fact that you use mtex > 6 for what it was not written for (which is easy to adapt but ther version is still very beta, so I wouldn't use this mtex version for anything but testing finding bugs at the moment).
Cheers,
Rüdiger

**Reply by NikitaSakharov** [reply]
Hello, Rudiger!

Many many many thanks! MTEX v5.8 really works fine with loadEBSD_EDAXh5 you've provided!

**Reply by NikitaSakharov** [reply]
Hello, Rudiger.

I've worked a little with loadEBSD_EDAXh5 and found it does not work correctly with some data. It seems, that there are datasets in h5 file, which have several datasets inside theirselves. 

So I get an error like this.

Expected one output from a curly brace or dot indexing expression, but there were 2 results.

Error in loadEBSD_EDAXh5 (line 16)
ebsddatapath = ci5.Groups.Groups.Groups(dset).Groups.Groups(3).Groups(1).Groups(1);

Error in ebsdListEDAX2 (line 12)
    ebsd = loadEBSD_EDAXh5('D:\myPath\myH5File.h5','dataset',1)
